dpp/src/state_transition/state_transition_factory.rs

```diff
@@ -0,0 +1,236 @@
+use anyhow::anyhow;
+use std::convert::TryFrom;
+
+use crate::{
+    consensus::{basic::BasicError, ConsensusError},
+    data_contract::{
+        state_transition::{DataContractCreateTransition, DataContractUpdateTransition},
+        DataContract,
+    },
+    document::DocumentsBatchTransition,
+    identity::state_transition::{
+        identity_create_transition::IdentityCreateTransition,
+        identity_credit_withdrawal_transition::IdentityCreditWithdrawalTransition,
+        identity_topup_transition::IdentityTopUpTransition,
+    },
+    prelude::Identifier,
+    state_repository::StateRepositoryLike,
+    util::json_value::JsonValueExt,
+    ProtocolError,
+};
+use serde_json::Value as JsonValue;
+
+use super::{StateTransition, StateTransitionType};
+
+pub async fn create_state_transition(
+    state_repository: &impl StateRepositoryLike,
+    raw_state_transition: JsonValue,
+) -> Result<StateTransition, ProtocolError> {
+    let transition_type = try_get_transition_type(&raw_state_transition)?;
+
+    if transition_type == StateTransitionType::DocumentsBatch {
+        let maybe_transitions = raw_state_transition
+            .get("transitions")
+            .ok_or_else(|| anyhow!("the transitions property doesn't exist"))?;
+        if let Some(transitions) = maybe_transitions.as_array() {
+            let data_contracts =
+                fetch_data_contracts_for_document_transition(state_repository, transitions).await?;
+            let documents_batch_transition =
+                DocumentsBatchTransition::from_raw_object(raw_state_transition, data_contracts)?;
+            return Ok(StateTransition::DocumentsBatch(documents_batch_transition));
+        }
+    }
+
+    match transition_type {
+        StateTransitionType::DataContractCreate => {
+            let transition = DataContractCreateTransition::from_raw_object(raw_state_transition)?;
+            Ok(StateTransition::DataContractCreate(transition))
+        }
+        StateTransitionType::DataContractUpdate => {
+            let transition = DataContractUpdateTransition::from_raw_object(raw_state_transition)?;
+            Ok(StateTransition::DataContractUpdate(transition))
+        }
+        StateTransitionType::IdentityCreate => {
+            let transition = IdentityCreateTransition::new(raw_state_transition)?;
+            Ok(StateTransition::IdentityCreate(transition))
+        }
+        StateTransitionType::IdentityTopUp => {
+            let transition = IdentityTopUpTransition::new(raw_state_transition)?;
+            Ok(StateTransition::IdentityTopUp(transition))
+        }
+        StateTransitionType::IdentityCreditWithdrawal => {
+            let transition =
+                IdentityCreditWithdrawalTransition::from_raw_object(raw_state_transition)?;
+            Ok(StateTransition::IdentityCreditWithdrawal(transition))
+        }
+        StateTransitionType::DocumentsBatch => {
```

Comment:
Isn't that the same thing as in the `if` block in the beginning of the function?

Reply:
yes it is! I forgot to remove the first one after refactor, thanks!